core/consensus/src/consensus.rs

```diff
@@ -57,13 +62,84 @@ impl<Adapter: ConsensusAdapter + 'static> Consensus for OverlordConsensus<Adapte
         Ok(())
     }
 
-    async fn update_epoch(
-        &self,
-        _ctx: Context,
-        _epoch: Epoch,
-        _signed_txs: Vec<SignedTransaction>,
-        _proof: Proof,
-    ) -> ProtocolResult<()> {
+    async fn update_epoch(&self, ctx: Context, msg: Vec<u8>) -> ProtocolResult<()> {
+        // Reveive the rich epoch ID.
+        let epoch_id: FixedEpochID =
+            deserialize(&msg).map_err(|_| ConsensusError::DecodeErr(MsgType::RichEpochID))?;
+        let mut current_epoch_id = self.engine.get_current_epoch_id(ctx.clone()).await?;
+        let rich_epoch_id = epoch_id.inner;
+
+        if current_epoch_id >= rich_epoch_id {
+            return Ok(());
+        } else if current_epoch_id == rich_epoch_id - 1 {
+            // Wait for an interval and re-check if needs to synchronization.
+            Delay::new(Duration::from_millis(BROADCAST_STATUS_INTERVAL))
+                .await
+                .map_err(|err| ConsensusError::Other(format!("{:?}", err)))?;
+
+            current_epoch_id = self.engine.get_current_epoch_id(ctx.clone()).await?;
+            if rich_epoch_id <= current_epoch_id {
+                return Ok(());
+            }
+        }
+
+        let mut prev_hash = self
+            .engine
+            .get_epoch_by_id(ctx.clone(), current_epoch_id + 1)
+            .await?
+            .header
+            .pre_hash;
+
+        // Check epoch for the first time.
+        let epoch_header = self
+            .engine
+            .pull_epoch(ctx.clone(), current_epoch_id + 1)
+            .await?
+            .header;
+        self.check_proof(current_epoch_id + 1, epoch_header.proof.clone())?;
+        if prev_hash != epoch_header.pre_hash {
+            return Err(ConsensusError::SyncEpochHashErr(current_epoch_id + 1).into());
+        }
+
+        // Lock the consensus engine, block commit process.
+        self.engine.lock.lock().await;
```

Comment:
This is a problem if you have multiple threads competing for locks at the same time.

First, these threads will get the lock-in turn, and then they will all execute the synchronization process once, although possibly repeatedly.